autoload/dirvish.vim

```diff
@@ -136,7 +136,7 @@ function! dirvish#shdo(paths, cmd) abort
       \.'|buffer '.bufnr('%').'|setlocal bufhidden=wipe|endif'
   augroup END
 
-  nnoremap <buffer><silent> Z! :silent write<Bar>exe '!'.(has('win32')?'':shellescape(&shell)).' %'<Bar>if !v:shell_error<Bar>close<Bar>endif<CR>
+  nnoremap <buffer><silent> Z! :silent write<Bar>exe '!'.(has('win32')?fnameescape(escape(expand('%:p:gs?\\?/?'), '&\')):shellescape(&shell).' %')<Bar>if !v:shell_error<Bar>close<Bar>endif<CR>
```

Comment:
Why not use `shellescape(fnamemodify('%:p'))` ? It should already take care of all of these details, that is the purpose of shellescape().

Reply:
Can you please clarify your suggested function as `fnamemodify` requires `{mod}`. Else, the following error messages are returned:
```
E119: Not enough arguments for function: fnamemodify
E116: Invalid arguments for function shellescape(fnamemodify('%:p')))
E15: Invalid expression: (shellescape(fnamemodify('%:p')))
```

Reply:
shellescape(fnamemodify(@%,':p'))

Reply:
No, that will not work with `Z!` executed via `sh.exe` (i.e. `git-bash`) on Windows OS.

![1](https://user-images.githubusercontent.com/13987638/56460301-f94cfc00-63a0-11e9-8c13-be6fc5ef2578.png)

As you can see in the attached pic, the following problem(s) arises:
* Since you did not convert `\` to `/` and `shellescape()` encloses the given filepath in double quotes, path separators will consequently get dropped as `bash` treats `\` as an escape char inside `""`.
* This also means that pairing commands with registers `%` and `#` will not work in `sh.exe` (e.g. `:!mkdir %foo`), but an easy fix would be, for instance, adding a command-line mode mapping (i.e. `cnoremap <expr> %% fnameescape(escape(expand('%:p:~:h:gs?\\?/?'), '&\')) . '/'`) into one's `_vimrc`. Then, one can simply `:!mkdir %%foo`.

I understand my solution may not be as polished due to my limited knowledge in `vimscript`, though, it is the only thing I could come up with thus far to work reliably across `sh.exe`, `cmd.exe` and `powershell.exe`. Please, let me know if you have further suggestions on improving the code.